Sources/ApiVideoLiveStream.swift

```diff
@@ -250,6 +277,17 @@ public class ApiVideoLiveStream {
     private func didEnterBackground(_: Notification) {
         stopStreaming()
     }
+
+    @objc
+    private func zoom(sender: UIPinchGestureRecognizer) {
+        if sender.state == .began || sender.state == .changed {
+            if(sender.scale < 2 && sender.scale > 0) {
+                currentZoomFactor = min(max(currentZoomFactor + (min(max(sender.scale, 0.5), 1.5) - 1) * 2.2, 0), 4)
```

Comment:
Have you read this line:
https://github.com/shogo4405/HaishinKit.swift/blob/64571447d123be0a92ab58562a0c183a3799e0db/Platforms/iOS/AVVideoIOUnit%2BExtension.swift#L15

I guess you don't have to set limit to 0 and 4. HK already does this.

Anyway, I am not sure to understand the formula. Can you explain?

Reply:
The outer min-max is a boundary for the total zoom factor. Yes, it is to an extend set by HK. But here we are setting a local variable, meaning HK does not clamp our currentZoomFactor. So having our own clamp as a measure for the user that keeps zooming. Without this clamp you can zoom to a factor of 100 or -100, meaning you will have to zoom all the way towards 0 again before anything is affected again. 

currentZoomFactor in the calculation is the previous zoom factor.

The inner min-max is a boundary for how much you can affect the zoom in a single frame. (This is just to ignore outlier data - It is fairly hard to go past 0.9 or 1.1, but it can happen that the user somehow tricks the phone into thinking you moved a finger from one corner to the other in a single frame) - This is not a necessary, can be removed.

`sender.scale`: is the ratio change in the gesture compared to the last frame. We minus this value by 1 in order to magnify the scale, and it makes it so we can just add it to the previous zoom factor value (currentZoomFactor)
The `* 2.2` is simply to magnify the zoom. If this wasnt present the zoom would feel very slow.

Simplified: 
We take the pinch scale minus 1 (Above 0 == zoom in | Below 0 == zoom out), we multiply this value by a multiplier to magnify the effect of the pinch action. Then we take that value and add it to the previous zoom factor.
Now this sum value is the new zoom factor which we save in currentZoomFactor & apply this to HK zoom factor.
Afterwards we reset the `sender.scale` to 1 so it is ready for the next frame.

I did try to set sender.scale = 0, but that just stops the pinch event.

Reply:
The ramping & withRate. Did try them out, and the zooming felt slow. But could make a zoomConfig that includes the options ramp, withRate, pinchZoomMultiplier & maxZoom.

Reply:
The only thing that trigger me, is that if the device has more a maximum zoom > 4, the user won't be able to access his maximum zoom. Moreover, the minimum zoom is 1.0, so when the `currentZoomFactor` is going from 0 to 1, nothing happens for the user till the zoom factor is above 1.0.

A way to deal with this, could be to not clamp the `currentZoomFactor` and then use HK `zoomFactor` API (https://github.com/shogo4405/HaishinKit.swift/blob/64571447d123be0a92ab58562a0c183a3799e0db/Platforms/iOS/AVVideoIOUnit%2BExtension.swift#L7) like this:
```
let zoomFactor = currentZoomFactor + (sender.scale-1) * pinchZoomMultiplier
rtmpStream.setZoomFactor(currentZoomFactor)
currentZoomFactor = rtmpStream.zoomFactor
```

The other way is to clamp with the real value from `1.0` to `device.activeFormat.videoMaxZoomFactor`